### ISSUE TITLE
Type checking for pattern matching against optionals

### ISSUE THREAD

**Issue by jeromesimeon** [accordproject/ergo#502]
It's unclear that any of those should be allowed:
```
$ return match 1 with let? x then 1 else 0
Response. 0 : Integer
$ return match 1 with let? x then x else 0
Response. 0 : Integer
$ return 1 ?? 0
Response. 0 : Integer
$ define concept A{x:String}
$ return match A{x:"foo"} with let? x: A then x else 0
Response. 0 : Any
$ return 1?.a
Response. none : Nothing?
$ return {a: 1}?.a
Response. none : Nothing?
```

Although one could argue that this is a reasonable semantics, I think one could make a case that pattern matching should work on types consistent with the pattern (`with let x?` only for option types, `with let x : B` only for classes, `with let x = 1` only for Integer, etc.).

Note that if considering `e1 ?? e2` to be syntactic sugar / consistent with pattern matching, then issue #387 argue that optional matching against a non-optional type should be an error.

**Comment by jeromesimeon** [comment]
Proposed resolution is to change the type checker for `let? x` pattern matching to raise a type error for non-optional types.

**Comment by jeromesimeon** [comment]
Addressed in #503
